src/riscv/lib/src/machine_state/memory.rs

```diff
@@ -313,6 +339,47 @@ pub trait MemoryConfig: Send + 'static {
         F: FnManager<Ref<'a, M>>;
 }
 
+/// State of a range of memory pages
+pub enum PageState {
+    /// The range is immutable and marked as executable.
+    ImmutableExec,
+
+    /// No beneficial properties hold for the range.
+    Other,
+}
+
+impl From<Permissions> for PageState {
+    fn from(permissions: Permissions) -> Self {
+        if permissions.can_read() && permissions.can_exec() && !permissions.can_write() {
+            Self::ImmutableExec
+        } else {
+            Self::Other
+        }
+    }
+}
+
+/// Implementing types will be informed about changes to memory permissions
+pub trait MemoryGovernanceListener {
+    /// Called on changes to memory permissions.
+    fn handle_memory_permission_change(&mut self, pages: RangeInclusive<usize>, state: PageState);
+}
+
+/// Default memory governance handler that does nothing
+#[cfg(test)]
+#[derive(Default)]
+pub struct DummyMemoryGovernanceHandler;
```

Comment:
Why would we want to have tests use a dummy handler that is a no-op for memory governance. Especially since you added the test-only method `set_all_readable_writeable` which should be used in tests to make memory handling easier. You either want diverging behaviour with a dummy handler or introduce helper functions to help use the actual types used in production, I don't see why you would have both.

In tests, probably what you want to do is, depending if you are using the jit or not, to pass the proper handler. Which currently should be no-ops.

Reply:
From offline: This is needed because some tests use a lower-level API (`MachineState` vs `MachineCoreState`), so it make sense to have this one, at least for now probably it's not worth the effort of getting rid of this dummy.